File: C++/Medium/WrongAnswer/split-a-string-into-the-max-number-of-unique-substrings_9.cpp
class Solution {
public:
    int maxUniqueSplit(string s) {
        
        int n = s.size();
        vector<int> dp(n+1, 0);
        for (int i = 1; i <= n; ++i) {
            unordered_set<string> st;
            bool flag = true;
            for (int j = i; j >= 1; --j) {
                string sub = s.substr(j-1, i-j+1);
                if (st.count(sub)) {
                    flag = false;
                    break;
                }
                st.insert(sub);
            }
            if (flag) dp[i] = st.size();
            else dp[i] = 0;
            for (int j = 1; j < i; ++j) {
                dp[i] = max(dp[i], dp[j] + dp[i-j]);
            }
        }
        return dp[n];
    }
};

File: Java/Medium/Accepted/split-a-string-into-the-max-number-of-unique-substrings_9.java
class Solution {
    public int maxUniqueSplit(String s) {
        
        return split(s, new HashSet<String>());
    }
    
    private int split(String s, Set<String> set) {
        if (s.length() == 0) {
            return 0;
        }
        
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            String sub = s.substring(0, i + 1);
            if (!set.contains(sub)) {
                set.add(sub);
                max = Math.max(max, 1 + split(s.substring(i + 1), set));
                set.remove(sub);
            }
        }
        return max;
    }
}